oximeter/collector/src/agent.rs

```diff
@@ -131,29 +125,153 @@ async fn perform_collection(
                 "failed to send collection request to producer";
                 "error" => ?e
             );
-            self_target
-                .failures_for_reason(self_stats::FailureReason::Unreachable)
-                .datum
-                .increment()
+            Err(self_stats::FailureReason::Unreachable)
         }
     }
 }
 
+// The type of one collection task run to completion.
+//
+// An `Err(_)` means we failed to collect, and contains the reason so that we
+// can bump the self-stat counter accordingly.
+type CollectionResult = Result<ProducerResults, self_stats::FailureReason>;
+
+// The return type of a spawned task running `perform_collection`.
+type CollectionTaskResult = JoinHandle<CollectionResult>;
+
+// An optional future we can poll in an outstanding collection.
+type CollectionFuture = Pin<Box<CollectionTaskResult>>;
+
+/// A outstanding task running `perform_collection`.
+struct OutstandingCollection {
+    /// The future resolving to the collection task result, if any.
+    fut: Option<CollectionFuture>,
+    /// The token we were provided if this was a triggered collection.
+    token: Option<CollectionToken>,
+    /// Logger to run each collection with.
+    log: Logger,
+    /// Client used to run each collection with.
+    client: reqwest::Client,
+}
+
+impl core::future::Future for OutstandingCollection {
+    type Output = Result<
+        (
+            Result<ProducerResults, self_stats::FailureReason>,
+            Option<CollectionToken>,
+        ),
+        JoinError,
+    >;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Self::Output> {
+        match self.fut.as_mut() {
+            Some(handle) => {
+                match handle.as_mut().poll(cx) {
+                    Poll::Ready(res) => {
+                        // NOTE: Take out of the handle too.
+                        //
+                        // This is important. `JoinHandle` will panic if we poll
+                        // it after completion, so remove it here so that we
+                        // never enter this match arm at all if we have no
+                        // handle, on the next time we're polled.
```

Comment:
I agree that this is awkward, but it's not really a "single future being ready multiple times". This is similar to [`OptionFuture`](https://docs.rs/futures/latest/futures/future/struct.OptionFuture.html), but probably less well-thought out. I'm happy to go a different direction, using some channels to control a spawned collection rather than this method.